core/src/main/java/org/apache/iceberg/rest/auth/AuthSessionCache.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.rest.auth;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import java.time.Duration;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ForkJoinPool;
+import java.util.function.Function;
+import java.util.function.LongSupplier;
+import javax.annotation.Nullable;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+
+/** A cache for {@link AuthSession} instances. */
+public class AuthSessionCache implements AutoCloseable {
+
+  private final Duration sessionTimeout;
+  private final Executor executor;
+  private final LongSupplier nanoTimeSupplier;
+
+  private volatile Cache<String, AuthSession> sessionCache;
+
+  /**
+   * Creates a new cache with the given session timeout, and with default executor and nano time
+   * supplier for eviction tasks.
+   *
+   * @param sessionTimeout the session timeout. Sessions will become eligible for eviction after
+   *     this duration of inactivity.
+   */
+  public AuthSessionCache(Duration sessionTimeout) {
```

Comment:
It looks like we should also make this package private.  The other issue is that we shouldn't be using the common pool by default.  This should probably be provided by the `OAuth2Manager` or if we expect there to be only one per manager, then just create a named pool for handling refreshes.   We definitely don't want to rely on or possibly tie up the common pool.

Reply:
@danielcweeks we are already using the common pool. The session cache currently is created as follows:

https://github.com/apache/iceberg/blob/a2b8008da7bc26e03248a35eeee60d1cc7e8499d/core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java#L1144-L1154

As you can see, we are not providing an executor explicitly, so evictions are being done on the common pool.

Reply:
Re: turning this constructor package-private: this won't fly, as SigV4 will also need to create caches, so this constructor needs to be public.

Reply:
@danielcweeks I searched for all `Caffeine` caches in Iceberg's repo and found 25 occurrences. Only one occurrence sets the executor, the one in `CachingCatalog`:

https://github.com/apache/iceberg/blob/f7d40f07131894719c7ad9f99fda7c6cb7aec932/core/src/main/java/org/apache/iceberg/CachingCatalog.java#L112

Consequently, all the others are using the common pool for asynchronous tasks such as eviction.

Reply:
Yes, we previously had refresh in a [dedicated threadpool](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java#L730), so I don't think the other caffeine examples are what we should point to.

Also, we don't want to rely on the common pool for something important like token refresh.  If something ties up the thread pool, we may miss the window to refresh.  Also, we don't want to tie up the thread pool with refresh threads if something goes wrong.

Reply:
This is not about token refresh, this is about auth session eviction.

But fair enough. I'll do it.